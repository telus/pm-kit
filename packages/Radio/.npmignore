*.stories.{js,mdx}
rollup.config.js
Radio.jsx
__tests__/