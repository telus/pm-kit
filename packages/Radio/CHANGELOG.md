# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 1.0.0 (2020-01-22)


### Features

* **radio:** added Radio package ([#30](https://github.com/telus/pm-kit/issues/30)) ([22b3349](https://github.com/telus/pm-kit/commit/22b33493bb4884d611c982309ceb50b244db627c))
