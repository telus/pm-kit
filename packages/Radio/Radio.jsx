import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import { css } from '@emotion/core'
import generateId from '../../shared/utils/generateId/generateId.js'
import { parkGreen, softSandBrown, lilyGreen, lightTan } from '@pm-kit/colours'
import { motion } from 'framer-motion'

const radio = css`
  opacity: 0;
  position: fixed;
  pointer-events: none;
  &:checked ~ label > div {
    background-color: initial;
  }
  &:checked ~ label > div > span {
    display: block;
  }
`
const basicRadioFake = css`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 17px;
  height: 17px;
  min-height: 17px;
  min-width: 17px;
  margin-top: -2px;
  outline: 0;
  line-height: 0;
  border-radius: 50%;
  border: 2px solid ${parkGreen};
  cursor: pointer;
  transition: border-color 0.1s linear, background-color 0.1s linear;
`

const radioFakeSelected = css`
  border: 2px solid ${softSandBrown};
`

const radioInner = css`
  display: none;
  width: 9px;
  height: 9px;
  border-radius: 50%;
  background-color: ${softSandBrown};
`

const basicLabel = css`
  display: flex;
  align-items: center;
  padding: 12px 18px;
  border-radius: 12px;
  cursor: pointer;
  overflow: hidden;
  background-position: right bottom;
  background-color: ${parkGreen};
  width: 100%;
`

const labelNotSelected = css`
  &:hover {
    background-color: ${lilyGreen};
  }
  background-color: ${lightTan};
`

const labelSelected = css`
  :hover {
    background-color: #09585e;
  }
  background-color: ${parkGreen};
  color: ${softSandBrown};
`

const labelText = css`
  margin-left: 1rem;
  font-size: 1em;
  font-weight: 700;
`

export const Radio = ({
  checked,
  description,
  error,
  feedback,
  id,
  label,
  name,
  value,
  variant,
  forwardedRef,
  ...rest
}) => {
  const radioId = generateId(id, rest.name, label)
  const radioFakeStyle = [basicRadioFake]
  const labelStyle = [basicLabel, labelNotSelected]
  const radioVariant = {
    whileTap: {},
  }

  if (checked) {
    labelStyle.push(labelSelected)
    radioFakeStyle.push(radioFakeSelected)
  }

  const labelVariant = {
    whileTap: {
      scale: 0.96,
    },
  }

  return (
    <motion.div whileTap="whileTap" variants={radioVariant}>
      <input
        // aria-describedby={feedback === 'error' ? getErrorId(name, value, id) : undefined}
        aria-invalid={feedback === 'error'}
        checked={checked}
        css={radio}
        data-testid="hidden-input"
        id={radioId.identity()}
        name={name}
        // ref={ref}
        type="radio"
        value={value}
        {...rest}
      />
      <motion.label
        css={labelStyle}
        data-testid="checkbox-label"
        htmlFor={radioId.identity()}
        variants={labelVariant}
        transition={{ duration: 0.65 }}
      >
        <div css={radioFakeStyle} data-testid="fake-input">
          <span css={radioInner} />
        </div>
        <span css={labelText}>{label}</span>
      </motion.label>
    </motion.div>
  )
}

Radio.propTypes = {
  /**
   * The label.
   */
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * Associate this radio with a group. Set as the HTML name attribute.
   */
  name: PropTypes.string.isRequired,
  /**
   * Control the appearance of the component
   */
  variant: PropTypes.oneOf(['borderless', 'bordered']),
  /**
   * The value. Must be unique within the group.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]).isRequired,
  /**
   * Use `checked` for controlled radio. For uncontrolled radio, use React's built-in `defaultChecked` prop.
   * See examples below for more details.
   */
  checked: PropTypes.bool,
  /**
   * Description text below the radio.
   */
  description: PropTypes.string,
  /**
   * The id. Must be unique within the group. If no id is provided, one will be generated in the following format: `name_value`
   */
  id: PropTypes.string,
  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(['error']),
  /**
   * An error message.
   */
  error: PropTypes.string,
}

Radio.defaultProps = {
  variant: 'borderless',
  description: undefined,
  feedback: undefined,
  error: undefined,
  checked: undefined,
  label: '',
  name: '',
  value: '',
  id: '1',
}

Radio.displayName = 'Radio'

const RadioWithRef = forwardRef((props, ref) => <Radio {...props} forwardedRef={ref} />)

export default RadioWithRef
