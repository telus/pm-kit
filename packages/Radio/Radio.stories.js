import React, { useState } from 'react'
import { withKnobs, text } from '@storybook/addon-knobs'
import Radio from './Radio'

export default {
  title: 'Design System|Radio',
  component: Radio,
  decorators: [withKnobs],
}

export const Default = () => {
  const [clicked, setClicked] = useState('en')

  const handleSelectOption = en_ => {
    setClicked(en_)
  }

  return (
    <>
      <Radio
        checked={clicked === 'en'}
        label="English"
        name="prefLang1"
        onChange={() => {
          handleSelectOption('en')
        }}
        value={false}
      />
      <Radio
        id="2"
        checked={clicked === 'fr'}
        label="French"
        name="prefLang2"
        onChange={() => {
          handleSelectOption('fr')
        }}
        value={true}
      />
    </>
  )
}

export const Playground = () => {
  const [clicked, setClicked] = useState('en')

  const handleSelectOption = en_ => {
    setClicked(en_)
  }

  return (
    <>
      <Radio
        id="1"
        checked={clicked === 'fr'}
        label={text('Label', 'Testing Radio 1')}
        name="prefLang1"
        onChange={() => {
          handleSelectOption('fr')
        }}
        value={true}
      />
      <Radio
        id="2"
        checked={clicked === 'en'}
        label={text('Label', 'Testing Radio 2')}
        name="prefLang2"
        onChange={() => {
          handleSelectOption('en')
        }}
        value={true}
      />
    </>
  )
}

Playground.story = {
  name: 'playground',
  parameters: { docs: { page: null, disable: true } },
}
