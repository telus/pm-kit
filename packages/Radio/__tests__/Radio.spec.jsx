import React from 'react'
import { render } from 'enzyme'

import Radio from '../Radio'

describe('Radio', () => {
  it('renders', () => {
    const radio = render(<Radio label="Test1" />)

    expect(radio).toMatchSnapshot()
  })
})
