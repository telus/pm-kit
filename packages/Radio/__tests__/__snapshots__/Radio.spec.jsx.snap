// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders 1`] = `
.emotion-0 {
  opacity: 0;
  position: fixed;
  pointer-events: none;
}

.emotion-0:checked ~ label > div {
  background-color: initial;
}

.emotion-0:checked ~ label > div > span {
  display: block;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px 18px;
  border-radius: 12px;
  cursor: pointer;
  overflow: hidden;
  background-position: right bottom;
  background-color: #034045;
  width: 100%;
  background-color: #dbd7c9;
}

.emotion-4:hover {
  background-color: #B6EBDB;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 17px;
  height: 17px;
  min-height: 17px;
  min-width: 17px;
  margin-top: -2px;
  outline: 0;
  line-height: 0;
  border-radius: 50%;
  border: 2px solid #034045;
  cursor: pointer;
  -webkit-transition: border-color 0.1s linear,background-color 0.1s linear;
  transition: border-color 0.1s linear,background-color 0.1s linear;
}

.emotion-1 {
  display: none;
  width: 9px;
  height: 9px;
  border-radius: 50%;
  background-color: #f4e8d9;
}

.emotion-3 {
  margin-left: 1rem;
  font-size: 1em;
  font-weight: 700;
}

<div>
  <input
    aria-invalid="false"
    class="emotion-0"
    data-testid="hidden-input"
    id="1"
    name=""
    type="radio"
    value=""
  />
  <label
    class="emotion-4"
    data-testid="checkbox-label"
    for="1"
  >
    <div
      class="emotion-2"
      data-testid="fake-input"
    >
      <span
        class="emotion-1"
      />
    </div>
    <span
      class="emotion-3"
    />
  </label>
</div>
`;
